charts/portworx/templates/storage-cluster.yaml

```diff
@@ -1,193 +1,84 @@
-      - {{ $name }}
-      {{- end }}
-    {{- if and (ne $kvdbDevice "none") (not (hasPrefix "/" $kvdbDevice)) }}
-    kvdbDeviceSpec: {{ $kvdbDevice }}
-    {{- end }}
-    {{- if and (ne $journalDevice "none") (not (hasPrefix "/" $journalDevice)) }}
-    journalDeviceSpec: {{ $journalDevice }}
-    {{- end }}
-    {{- if ne (.Values.maxStorageNodesPerZone | int) 0 }}
-    maxStorageNodesPerZone: {{ .Values.maxStorageNodesPerZone }}
-    {{- end }}
-{{- else }}
-  {{- if $AKSInstall }}
+  {{- with .Values.cloudStorage }}
   cloudStorage:
-    deviceSpecs:
-      - type=Premium_LRS,size=150
-    {{- if and (ne $kvdbDevice "none") (not (hasPrefix "/" $kvdbDevice)) }}
-    kvdbDeviceSpec: {{ $kvdbDevice }}
-    {{- end }}
-    {{- if and (ne $journalDevice "none") (not (hasPrefix "/" $journalDevice)) }}
-    journalDeviceSpec: {{ $journalDevice }}
-    {{- end }}
-    {{- if gt (.Values.maxStorageNodesPerZone | int) 0 }}
-    maxStorageNodesPerZone: {{ .Values.maxStorageNodesPerZone }}
-    {{- end }}
-  {{- else }}
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
+  {{- with .Values.storage }}
   storage:
-    {{- if hasPrefix "/" $drives }}
-    devices:
-      {{- $driveNames := $drives | split ";" }}
-        {{- range $index, $name := $driveNames }}
-        - {{ $name }}
-        {{- end }}
-    {{- end }}
-    {{- if or $usefileSystemDrive $deployEnvironmentIKS }}
-    forceUseDisks: true
-    {{- end }}
-    {{- if eq $usedrivesAndPartitions true }}
-    useAllWithPartitions: true
-    {{- else }}
-    useAll: true
-    {{- end }}
-    {{- if and (ne $kvdbDevice "none") (hasPrefix "/" $kvdbDevice) }}
-    kvdbDevice: {{ $kvdbDevice }}
-    {{- end }}
-    {{- if and (ne $journalDevice "none") (or (hasPrefix "/" $journalDevice) (eq "auto" $journalDevice)) }}
-    journalDevice: {{ $journalDevice }}
-    {{- end }}
+    {{- toYaml . | nindent 4 }}
   {{- end }}
-{{- end }}
-
-  {{- if or (ne $dataInterface "none") (ne $managementInterface "none") }}
+    kvdbDeviceSpec: {{ .Values.kvdb.device }}
+    journalDeviceSpec: {{ .Values.journalDevice }}
+    maxStorageNodesPerZone: {{ .Values.maxStorageNodesPerZone }}
+    forceUseDisks: {{ .Values.forceUseDisks }}
+    useAllWithPartitions: {{ .Values.useAllWithPartitions }}
+    useAll: {{ .Values.useAll }}
+    kvdbDevice: {{ .Values.kvdb.device }}
+    journalDevice: {{ .Values.journalDevice }}
   network:
-    {{- if ne $dataInterface "none" }}
-    dataInterface: {{ $dataInterface }}
-    {{- end }}
-    {{- if ne $managementInterface "none" }}
-    mgmtInterface: {{ $managementInterface }}
-    {{- end }}
-  {{- end }}
-
-  {{- if ne $secretType "none" }}
-  secretsProvider: {{$secretType}}
-  {{- else }}
-    {{- if $deployEnvironmentIKS }}
-  secretsProvider: ibm-kp
-    {{- end }}
-  {{- end }}
-
+    dataInterface: {{ .Values.network.dataInterface }}
+    mgmtInterface: {{ .Values.network.mgmtInterface }}
+  secretsProvider: {{ .Values.secretsProvider }}
```

Comment:
> These should only be present when the user passes in the value, else they should not exist in the spec.

This is a good point. But, what would happen if they all have a value of `null`? This is where we need to consult the developers on allowing the key to be set with a value of `null`, otherwise we're making the helm chart more complicated than it needs to be

Reply:
When the portworx operator parses a value as `null`, it will be populated as the zero-value in Go for that field type. So, if the result of `{{ .Values.network.dataInterface }}` is `null`, we are okay. 

If the value is `none`, that may cause problems.

With secretsProvider for example, if secretsProvider is `null`, we default to k8s type:
https://github.com/libopenstorage/operator/blob/adb782e67fb02bd1decc40f4e8e818e06b36eada/drivers/storage/portworx/portworx.go#L930-L932

This is the case for most fields. When `null` is set in other cases, fields will also behave the same as if they were unset altogether. I tested this:

```
// set to null, StorageCluster set to k8s
grant@grant-dev-vm:~/workspace/kubeup/test/stc$ cat stc.yaml  | grep secretsProvider
  secretsProvider: null
grant@grant-dev-vm:~/workspace/kubeup/test/stc$ kn get stc -o yaml | grep secretsProvider
    secretsProvider: k8s

// set to null, no value in StorageCluster set
grant@grant-dev-vm:~/workspace/kubeup/test/stc$ cat stc.yaml | grep dataInterface
    dataInterface: null
grant@grant-dev-vm:~/workspace/kubeup/test/stc$ kn get stc -o yaml | grep dataInterface    
```

Reply:
I would also like @piyush-nimbalkar to weigh in here, this will add all these options with `null` values in the STC and are we ok with that?

Reply:
The null values will only be part of the helm template. Once the object is created in K8s API server, it should get replaced by the correct default value in Go. Thus a yaml `null` value should be fine.